app/models/node.rb

```diff
@@ -102,11 +185,183 @@ class << self
+          m.on recursive_table[:id].eq(hierarchy[:parent_id])
+        else
+          m.on recursive_table[:parent_id].eq(hierarchy[:id])
+        end
+      end
+
+      union = non_recursive_term.union :all, recursive_term
+      as_statement = Arel::Nodes::As.new hierarchy, union
+
+      manager = select_manager.dup.tap do |m|
+        m.with :recursive, as_statement
+        m.from hierarchy
+        m.project hierarchy[Arel.star]
+      end
+    end
+
+    def all_descendants_via_arel_mgr
+      relatives_via_arel_mgr(false)
+    end
+
+    def all_descendants_via_arel
+      find_by_sql all_descendants_via_arel_mgr
+    end
+
+    def all_ancestors_via_arel_mgr
+      relatives_via_arel_mgr(true)
+    end
+
+    def all_ancestors_via_arel
+      find_by_sql all_ancestors_via_arel_mgr
+    end
+
+    # returns (...node, td_tag, ut_tag)
+    def node_system_tag_combos
+      n = Node.table
+      sts = Arel::Table.new :system_tags
+      n
+        .join(sts)
+        .on(n[:id].eq(sts[:anchored_id]).and(sts[:anchored_type].eq(Node.name)))
+        .with(TagDecl.system_tags.as(:system_tags.to_s))
+        .project(n[Arel.star], sts[:td_tag], sts[:ut_tag])
+    end
+
+    def all_node_system_tag_combos
+      find_by_sql node_system_tag_combos
+    end
+
+    # returns (base_id, ...node, td_tag, ut_tag)
+    def all_node_authz_reads
+      nwa = Arel::Table.new :nwa
+      nstc = Arel::Table.new :nstc
+
+      Arel::SelectManager.new
+        .project(nwa[Arel::star], nstc[:td_tag], nstc[:ut_tag])
+        .from(all_ancestors_via_arel_mgr.as("nwa"))
+        .join(node_system_tag_combos.as("nstc"))
+        .on(nstc[:id].eq(nwa[:id]))
+        .where(nstc[:td_tag].eq(Authz.read))
+    end
+
+    def anar_closest_parent
+      anar = Arel::Table.new :anar
+
+      Arel::SelectManager.new
+        .project(anar[:base_id], anar[:id].maximum.as("closest_permission_id"))
+        .from(all_node_authz_reads.as("anar"))
+        .group(anar[:base_id])
+    end
+
+    def with_permissioned_parent
+      acp = Arel::Table.new :acp
+      table
+        .join(anar_closest_parent.as("acp"))
+        .on(table[:id].eq(acp[:base_id]))
+        .project(table[Arel.star], acp[:closest_permission_id])
+    end
+
+    def old_node_authz_read
+      n = table
+      acp = Arel::Table.new :acp
+      anar = Arel::Table.new :anar
+
+      Arel::SelectManager.new
+        .project(acp[:base_id], anar[:ut_tag].as("group_name"))
+        .from(anar_closest_parent.as("acp"))
+        .join(all_node_authz_reads.as("anar"))
+        .on(anar[:id].eq(acp[:closest_permission_id]))
+    end
+
+    def old_nodes_readable_by(user)
+      nar = Arel::Table.new :nar
+      Arel::SelectManager.new
+        .from(old_node_authz_read.as("nar"))
+        .where(nar[:group_name].eq("all").or(nar[:group_name].in(user&.groups_arel)))
+        .project(nar[Arel.star])
+    end
+
+    def node_authz_read
```

Comment:
for whatever reason, joining with node_system_tag_combos is much more efficient in postgres than joining with all_node_authz_reads